conans/client/tools/net.py

```
@@ -60,21 +86,36 @@ def download(url, filename, verify=True, out=None, retry=None, retry_wait=None,
         downloader = CachedFileDownloader(config.download_cache, downloader, user_download=True)
-        downloader.download(url, filename, retry=retry, retry_wait=retry_wait, overwrite=overwrite,
-                            auth=auth, headers=headers, md5=md5, sha1=sha1, sha256=sha256)
-    else:
-        downloader.download(url, filename, retry=retry, retry_wait=retry_wait, overwrite=overwrite,
-                            auth=auth, headers=headers)
-        if md5:
-            check_md5(filename, md5)
-        if sha1:
-            check_sha1(filename, sha1)
-        if sha256:
-            check_sha256(filename, sha256)
 
-    out.writeln("")
+    def _download_file(downloader, url):
+        # The download cache is only used if a checksum is provided, otherwise, a normal download
+        if isinstance(downloader, CachedFileDownloader):
+            downloader.download(url, filename, retry=retry, retry_wait=retry_wait,
+                                overwrite=overwrite, auth=auth, headers=headers, md5=md5,
+                                sha1=sha1, sha256=sha256)
+        else:
+            downloader.download(url, filename, retry=retry, retry_wait=retry_wait,
+                                overwrite=overwrite, auth=auth, headers=headers)
+            if md5:
+                check_md5(filename, md5)
+            if sha1:
+                check_sha1(filename, sha1)
+            if sha256:
+                check_sha256(filename, sha256)
+        out.writeln("")
+
+    for index, url_it in enumerate(url):
+        try:
+            _download_file(downloader, url_it)
+            break
+        except (ConanConnectionError, NotFoundException, ConanException) as error:
+            out.warn("Could not download from the url {}.".format(url_it))
+            if (index + 1) == len(url):
+                raise ConanException("All {} URLs have failed: {}.".format(len(url),
+                                                                          str(error)))
```

This is the typical use-case for a `for/else` loop,  but I'm not sure if we want that logic in the codebase.

---

I think every failed URL should output the error message too. And the last one, maybe just raise the exception it got?

---

Almost it, but you point makes more sense. For now it raises a warning message, but I'll raise an error message, and a warning message the the retry.